## 1. Importing the necessary Python Modules / Libraries

In [4]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns
import plotly.express as px
import plotly.graph_objects as go

## 2. Reading the data set using the 'pandas' module

In [5]:
df = pd.read_excel('/content/drive/MyDrive/Final Lead Data.xlsx')

In [6]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [7]:
df.head()

ID First Name              Email Gender City                 Created  \
0  68112     ANIKET     aniket@xyz.com    NaN  NaN  04/27/2022 01:41:38 pm   
1  68110  Dhanshree  dhanshree@xyz.com    NaN  NaN  04/22/2022 04:08:38 pm   
2  68108     Dhiraj     dhiraj@xyz.com    NaN  NaN  04/16/2022 10:31:59 pm   
3  68106      Pooja      pooja@xyz.com    NaN  NaN  04/13/2022 10:05:15 pm   
4  68090     Aayush     aayush@xyz.com    NaN  NaN  03/26/2022 07:02:48 pm   

  Position         New College Name Colleges  Academic Year  \
0      NaN                      NaN      NaN            NaN   
1      NaN  Lords Universal College      NaN            NaN   
2      NaN                      NaN      NaN            NaN   
3      NaN                      NaN      NaN            NaN   
4      NaN        B.k Birla college      NaN            NaN   

  Branch/ Specialisation Other Branch What is your current academic year?  \
0                    NaN          NaN                                 NaN   
1                    NaN          NaN                                 NaN   
2                    NaN          NaN                                 NaN   
3                    NaN          NaN                                 NaN   
4                    NaN          NaN                                 NaN   

  Company Name/ College Name  \
0                        NaN   
1                        NaN   
2                        NaN   
3                        NaN   
4                        NaN   

  Would you like to know more about us and our programs?  \
0                                                yes       
1                                                yes       
2                                                yes       
3                                                yes       
4                                                yes       

  Are you interested in knowing more about our events?  \
0                                                yes     
1                                                yes     
2                                                yes     
3                                                yes     
4                                                yes     

  Have you recommended Cloud Counselage to anyone?  \
0                                               no   
1                                               no   
2                                               no   
3                                               no   
4                                               no   

  How did you come to know about this event?  
0                                        NaN  
1                                        NaN  
2                                        NaN  
3                                        NaN  
4                                        NaN

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5303 entries, 0 to 5302
Data columns (total 18 columns):
 #   Column                                                  Non-Null Count  Dtype  
---  ------                                                  --------------  -----  
 0   ID                                                      5303 non-null   int64  
 1   First Name                                              5303 non-null   object 
 2   Email                                                   5303 non-null   object 
 3   Gender                                                  200 non-null    object 
 4   City                                                    55 non-null     object 
 5   Created                                                 5303 non-null   object 
 6   Position                                                6 non-null      object 
 7   New College Name                                        1818 non-null   object 
 8   Colleges                              

## 3. Creating a copy of raw data so as to keep the main data as it as and doing further operations on the copy of the main data

In [9]:
final_df = df.copy()

In [10]:
final_df['Email'].value_counts()

Email
vaishnavi@xyz.com    37
shubham@xyz.com      35
yash@xyz.com         32
abhishek@xyz.com     29
sakshi@xyz.com       28
                     ..
julia@xyz.com         1
darrell@xyz.com       1
edie@xyz.com          1
kimberley@xyz.com     1
utkarsha@xyz.com      1
Name: count, Length: 2808, dtype: int64

## 4. Dropping duplicates if any using the 'Email' column from the dataset

In [11]:
final_df = final_df.drop_duplicates(subset='Email', keep='first')

In [12]:
final_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2808 entries, 0 to 5260
Data columns (total 18 columns):
 #   Column                                                  Non-Null Count  Dtype  
---  ------                                                  --------------  -----  
 0   ID                                                      2808 non-null   int64  
 1   First Name                                              2808 non-null   object 
 2   Email                                                   2808 non-null   object 
 3   Gender                                                  115 non-null    object 
 4   City                                                    54 non-null     object 
 5   Created                                                 2808 non-null   object 
 6   Position                                                1 non-null      object 
 7   New College Name                                        831 non-null    object 
 8   Colleges                                   

#### Let's drop the 'Position' column from the dataset as it contains only 1 non-null (Non-Blank) value

In [13]:
final_df = final_df.drop('Position', axis=1)

In [14]:
final_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2808 entries, 0 to 5260
Data columns (total 17 columns):
 #   Column                                                  Non-Null Count  Dtype  
---  ------                                                  --------------  -----  
 0   ID                                                      2808 non-null   int64  
 1   First Name                                              2808 non-null   object 
 2   Email                                                   2808 non-null   object 
 3   Gender                                                  115 non-null    object 
 4   City                                                    54 non-null     object 
 5   Created                                                 2808 non-null   object 
 6   New College Name                                        831 non-null    object 
 7   Colleges                                                662 non-null    object 
 8   Academic Year                              

### So, there are 2808 unique rows in the dataset provided

In [15]:
final_df['New College Name'].value_counts()

New College Name
Government Polytechnic Mumbai                                                    20
Datta Meghe College of Engineering                                               11
Sri Manakula Vinayagar Engineering College                                        9
Government polytechnic mumbai                                                     9
St. Francis Institute of Technology                                               8
                                                                                 ..
D.Y. Patil Institute of Master of Computer Applications and Management            1
Dr. D.Y. Patil Institute of MCA and Management                                    1
Dy Patil Institute of MCA and Management Akurdi pune                              1
D Y Patil Institute of Master of Computer Applications and Management, Akurdi     1
Don Bosco College Of Engineering Goa                                              1
Name: count, Length: 682, dtype: int64

In [16]:
final_df['Colleges'].value_counts()

Colleges
Others                                                                          168
Maharashtra Institute of Technology - MIT                                        18
Pimpri Chinchwad College of Engineering - PCCE                                   16
St. Francis Institute of Technology                                              15
KJ Somaiya College of Engineering                                                13
                                                                               ... 
Shrama Sadhana Bombay Trust's College of Engineering and Technology - SSCOET      1
North Maharashtra University                                                      1
Marathwada Mitra Mandal's Institute of Technology - MMIT                          1
Amrutvahini College of Engineering                                                1
Smt. Indira Gandhi College of Engineering - SIGCE                                 1
Name: count, Length: 165, dtype: int64

In [17]:
final_df['Branch/ Specialisation'].value_counts()

Branch/ Specialisation
Computer Science               473
Information Technology (IT)    307
Other                          305
Name: count, dtype: int64

In [18]:
final_df['Other Branch'].value_counts()

Other Branch
Electronics and telecommunication              28
Electronics and Telecommunication              17
Mechanical                                     14
Electronics                                    11
Mechanical Engineering                         11
                                               ..
information security and cloud computing        1
Electronics & Telecommunication Engineering     1
Electronic and Telecommunications               1
Mechanical Engg                                 1
Electrical & Electronics                        1
Name: count, Length: 131, dtype: int64

#### Let's combine the 'Branch / Specialisation' and 'Other Branch' columns to a new column named 'Branch' and then drop these two columns from the dataset

In [19]:
final_df['Branch'] = final_df['Branch/ Specialisation'] + ' ' + final_df['Other Branch']

In [20]:
columns = ['Branch/ Specialisation','Other Branch']
final_df = final_df.drop(columns, axis=1)

#### Let's combine the 'New College Name' and 'Colleges' columns to a new column named 'College' and then drop these two columns from the dataset

In [21]:
final_df['College'] = final_df['Colleges'] + ' ' + final_df['New College Name']

In [22]:
columns = ['Colleges','New College Name']
final_df = final_df.drop(columns, axis=1)

In [23]:
final_df.head()

ID First Name              Email Gender City                 Created  \
0  68112     ANIKET     aniket@xyz.com    NaN  NaN  04/27/2022 01:41:38 pm   
1  68110  Dhanshree  dhanshree@xyz.com    NaN  NaN  04/22/2022 04:08:38 pm   
2  68108     Dhiraj     dhiraj@xyz.com    NaN  NaN  04/16/2022 10:31:59 pm   
3  68106      Pooja      pooja@xyz.com    NaN  NaN  04/13/2022 10:05:15 pm   
4  68090     Aayush     aayush@xyz.com    NaN  NaN  03/26/2022 07:02:48 pm   

   Academic Year What is your current academic year?  \
0            NaN                                 NaN   
1            NaN                                 NaN   
2            NaN                                 NaN   
3            NaN                                 NaN   
4            NaN                                 NaN   

  Company Name/ College Name  \
0                        NaN   
1                        NaN   
2                        NaN   
3                        NaN   
4                        NaN   

  Would you like to know more about us and our programs?  \
0                                                yes       
1                                                yes       
2                                                yes       
3                                                yes       
4                                                yes       

  Are you interested in knowing more about our events?  \
0                                                yes     
1                                                yes     
2                                                yes     
3                                                yes     
4                                                yes     

  Have you recommended Cloud Counselage to anyone?  \
0                                               no   
1                                               no   
2                                               no   
3                                               no   
4                                               no   

  How did you come to know about this event? Branch College  
0                                        NaN    NaN     NaN  
1                                        NaN    NaN     NaN  
2                                        NaN    NaN     NaN  
3                                        NaN    NaN     NaN  
4                                        NaN    NaN     NaN

## 5. Let's Analyze the given data using Data Visualization modules/Libraries

In [24]:
gender = final_df['Gender'].value_counts().keys()
count = final_df['Gender'].value_counts().values
labels = [str(val) for val in count]

gender_plot_df = pd.DataFrame({
    'Gender': gender,
    'Count': count,
})

fig = px.bar(gender_plot_df, x='Gender', y='Count', text=labels, title='Count of Attendee by Gender', template='plotly',color='Gender', color_discrete_sequence=px.colors.qualitative.Plotly)

# Set the labels for the y-axis and x-axis
fig.update_yaxes(title_text='count')
fig.update_xaxes(title_text='Gender')

# Show the plot
fig.show()

In [25]:
city = final_df['City'].value_counts().keys()
count = final_df['City'].value_counts().values
labels = [str(val) for val in count]

city_plot_df = pd.DataFrame({
    'City': city,
    'Count': count,
})

fig = px.bar(city_plot_df, x='City', y='Count', text=labels, title='Count of Attendee by City', template='plotly',color='City', color_discrete_sequence=px.colors.qualitative.Plotly)

# Set the labels for the y-axis and x-axis
fig.update_yaxes(title_text='count')
fig.update_xaxes(title_text='City')

# Show the plot
fig.show()

In [26]:
branch = final_df['Branch'].value_counts().keys()
count = final_df['Branch'].value_counts().values
labels = [str(val) for val in count]

branch_plot_df = pd.DataFrame({
    'Branch': branch,
    'Count': count,
})

fig = px.bar(branch_plot_df, y='Branch', x='Count', text=labels, title='Count of Attendee by Branch', template='plotly',color='Branch', color_discrete_sequence=px.colors.qualitative.Plotly)

# Set the labels for the y-axis and x-axis
fig.update_xaxes(title_text='count')
fig.update_yaxes(title_text='Branch')


fig.update_layout(width=1100, height=2600)

# Show the plot
fig.show()

In [27]:
more_info = final_df['Would you like to know more about us and our programs?'].value_counts().keys()
count = final_df['Would you like to know more about us and our programs?'].value_counts().values
labels = [str(val) for val in count]

more_info_plot_df = pd.DataFrame({
    'More_info': more_info,
    'Count': count,
})

fig = px.bar(more_info_plot_df, x='More_info', y='Count', text=labels, title='Count of Attendee by which whether they want to know more about Cloud Counselage and their programs', template='plotly',color='More_info', color_discrete_sequence=px.colors.qualitative.Plotly)

# Set the labels for the y-axis and x-axis
fig.update_yaxes(title_text='count')
fig.update_xaxes(title_text='More_info about Cloud Counselage Pvt. Ltd. and their Programs')

# Show the plot
fig.show()

In [28]:
more_events = final_df['Are you interested in knowing more about our events?'].value_counts().keys()
count = final_df['Are you interested in knowing more about our events?'].value_counts().values
labels = [str(val) for val in count]

more_events_plot_df = pd.DataFrame({
    'More_events': more_events,
    'Count': count,
})

fig = px.bar(more_events_plot_df, x='More_events', y='Count', text=labels, title='Count of Attendee by which whether they want to know more about Cloud Counselage events', template='plotly',color='More_events', color_discrete_sequence=px.colors.qualitative.Plotly)

# Set the labels for the y-axis and x-axis
fig.update_yaxes(title_text='count')
fig.update_xaxes(title_text='More_events about Cloud Counselage Pvt. Ltd. events')

# Show the plot
fig.show()

In [29]:
recommend = final_df['Have you recommended Cloud Counselage to anyone?'].value_counts().keys()
count = final_df['Have you recommended Cloud Counselage to anyone?'].value_counts().values
labels = [str(val) for val in count]

recommend_plot_df = pd.DataFrame({
    'Recommend': recommend,
    'Count': count,
})

fig = px.bar(recommend_plot_df, x='Recommend', y='Count', text=labels, title='Count of Attendee by which whether they have Recommended Cloud Counselage', template='plotly',color='Recommend', color_discrete_sequence=px.colors.qualitative.Plotly)

# Set the labels for the y-axis and x-axis
fig.update_yaxes(title_text='count')
fig.update_xaxes(title_text='Recommended Cloud Counselage Pvt. Ltd.')

# Show the plot
fig.show()

In [30]:
event_info = final_df['How did you come to know about this event?'].value_counts().keys()
count = final_df['How did you come to know about this event?'].value_counts().values
labels = [str(val) for val in count]

event_info_plot_df = pd.DataFrame({
    'Event_info': event_info,
    'Count': count,
})

fig = px.bar(event_info_plot_df, x='Event_info', y='Count', text=labels, title='Count of Attendee by Event Information Platform/Source', template='plotly',color='Event_info', color_discrete_sequence=px.colors.qualitative.Plotly)

# Set the labels for the y-axis and x-axis
fig.update_yaxes(title_text='count')
fig.update_xaxes(title_text='Event Information Platform/Source')

# Show the plot
fig.show()

## 6. Let's use the 'Created' , 'Academic Year' and 'What is your current academic year?' columns to predict the Graduation Year of the attendee

In [31]:
final_df['Academic Year'].value_counts()

Academic Year
3.0    512
2.0    295
4.0    214
1.0     62
Name: count, dtype: int64

In [32]:
final_df['What is your current academic year?'].value_counts()

What is your current academic year?
3rd Year      21
Final Year    19
2nd Year      10
1st Year       4
Name: count, dtype: int64

#### Replacing values in 'What is your current academic year?' column: -
#### <ul><li>3rd Year to 3</li><li>Final Year to 4</li><li>2nd Year to 2</li><li>1st Year to 1</li></ul>

In [33]:
final_df['What is your current academic year?'] = final_df['What is your current academic year?'].replace({
    '3rd Year': 3,
    'Final Year': 4,
    '2nd Year': 2,
    '1st Year': 1
})

#### Replacing null values to 0

In [34]:
final_df['What is your current academic year?'] = final_df['What is your current academic year?'].fillna(0)

In [35]:
final_df['What is your current academic year?'] = final_df['What is your current academic year?'].astype(int)

#### Replacing values in 'Academic Year' column to their int data type from their original float data type

##### Replacing null values to 0

In [36]:
final_df['Academic Year'] = final_df['Academic Year'].fillna(0)

In [37]:
final_df['Academic Year'] = final_df['Academic Year'].astype(int)

In [38]:
final_df.tail(10)

ID  First Name               Email Gender City  \
5237  26036      ANJANA      anjana@xyz.com    NaN  NaN   
5238  26032    Gajrotar    gajrotar@xyz.com    NaN  NaN   
5240  26028       Nihal       nihal@xyz.com    NaN  NaN   
5245  25960  Akshinsinh  akshinsinh@xyz.com    NaN  NaN   
5246  25956      Charmi      charmi@xyz.com    NaN  NaN   
5250  25940        vraj        vraj@xyz.com    NaN  NaN   
5255  25920       Hemil       hemil@xyz.com    NaN  NaN   
5256  25918      Shitij      shitij@xyz.com    NaN  NaN   
5258  25914       dirgh       dirgh@xyz.com    NaN  NaN   
5260  25910    UTkarsha    utkarsha@xyz.com    NaN  NaN   

                     Created  Academic Year  \
5237  11/19/2019 10:59:56 am              0   
5238  11/18/2019 08:43:13 pm              0   
5240  11/17/2019 01:34:49 pm              0   
5245  10/25/2019 10:09:13 am              3   
5246  10/23/2019 07:06:38 pm              0   
5250  10/18/2019 01:03:15 pm              0   
5255  10/16/2019 01:37:18 pm              3   
5256  10/16/2019 01:35:50 pm              3   
5258  10/16/2019 11:28:11 am              0   
5260  10/16/2019 10:22:55 am              0   

      What is your current academic year? Company Name/ College Name  \
5237                                    0                        NaN   
5238                                    0                        NaN   
5240                                    0                        NaN   
5245                                    0                        NaN   
5246                                    0                        NaN   
5250                                    0                        NaN   
5255                                    0                        NaN   
5256                                    0                        NaN   
5258                                    0                        NaN   
5260                                    0                        NaN   

     Would you like to know more about us and our programs?  \
5237                                                 no       
5238                                                 no       
5240                                                 no       
5245                                                 no       
5246                                                 no       
5250                                                 no       
5255                                                 no       
5256                                                 no       
5258                                                 no       
5260                                                 no       

     Are you interested in knowing more about our events?  \
5237                                                 no     
5238                                                 no     
5240                                                 no     
5245                                                 no     
5246                                                 no     
5250                                                 no     
5255                                                 no     
5256                                                 no     
5258                                                 no     
5260                                                 no     

     Have you recommended Cloud Counselage to anyone?  \
5237                                               no   
5238                                               no   
5240                                               no   
5245                                               no   
5246                                               no   
5250                                               no   
5255                                               no   
5256                                               no   
5258                                               no   
5260                                               no   

     How did you come to know about this event? Branch

In [39]:
final_df['Academic Year'].value_counts()

Academic Year
0    1725
3     512
2     295
4     214
1      62
Name: count, dtype: int64

In [40]:
final_df['What is your current academic year?'].value_counts()

What is your current academic year?
0    2754
3      21
4      19
2      10
1       4
Name: count, dtype: int64

#### We have add values to 'Academic Year' column from 'What is your current academic year?' column where 'Academic Year' column was 0 (Null) and 'What is your current academic year?' column was non-zero (Non-Null)

In [41]:
final_df.loc[final_df['Academic Year'] == 0, 'Academic Year'] += final_df['What is your current academic year?']

In [42]:
final_df['Academic Year'].value_counts()

Academic Year
0    1710
3     515
2     297
4     221
1      65
Name: count, dtype: int64

#### Hence, we can now drop the 'What is your current academic year?' column from the dataset

In [43]:
final_df = final_df.drop('What is your current academic year?', axis=1)

#### Now, let's extract Month, Year from the 'Created' column of the dataset, using the datetime module

In [44]:
final_df['Created'] = pd.to_datetime(final_df['Created'])

<ipython-input-44-354308497516>:1: UserWarning:

Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.



In [45]:
final_df['Month'] = final_df['Created'].dt.month
final_df['Year'] = final_df['Created'].dt.year

In [46]:
final_df.head()

ID First Name              Email Gender City             Created  \
0  68112     ANIKET     aniket@xyz.com    NaN  NaN 2022-04-27 13:41:38   
1  68110  Dhanshree  dhanshree@xyz.com    NaN  NaN 2022-04-22 16:08:38   
2  68108     Dhiraj     dhiraj@xyz.com    NaN  NaN 2022-04-16 22:31:59   
3  68106      Pooja      pooja@xyz.com    NaN  NaN 2022-04-13 22:05:15   
4  68090     Aayush     aayush@xyz.com    NaN  NaN 2022-03-26 19:02:48   

   Academic Year Company Name/ College Name  \
0              0                        NaN   
1              0                        NaN   
2              0                        NaN   
3              0                        NaN   
4              0                        NaN   

  Would you like to know more about us and our programs?  \
0                                                yes       
1                                                yes       
2                                                yes       
3                                                yes       
4                                                yes       

  Are you interested in knowing more about our events?  \
0                                                yes     
1                                                yes     
2                                                yes     
3                                                yes     
4                                                yes     

  Have you recommended Cloud Counselage to anyone?  \
0                                               no   
1                                               no   
2                                               no   
3                                               no   
4                                               no   

  How did you come to know about this event? Branch College  Month  Year  
0                                        NaN    NaN     NaN      4  2022  
1                                        NaN    NaN     NaN      4  2022  
2                                        NaN    NaN     NaN      4  2022  
3                                        NaN    NaN     NaN      4  2022  
4                                        NaN    NaN     NaN      3  2022

### <p style='color:red'>Since the data was provided for Engineering colleges attendees so , they have 4 years of curriculum and 8 semesters.</p>

In [47]:
final_df['Year'] = final_df['Year'].astype(int)

In [48]:
year_of_grad = []
for i in range(len(final_df)):
    academic_year = final_df.iloc[i]['Academic Year']
    current_year = final_df.iloc[i]['Year']
    if academic_year != 0:
        if academic_year == 1:
            year_of_grad.append(3 + current_year)
        elif academic_year == 2:
            year_of_grad.append(2 + current_year)
        elif academic_year == 3:
            year_of_grad.append(1 + current_year)
        elif academic_year == 4:
            year_of_grad.append(current_year)
    else:
        year_of_grad.append(0)

In [49]:
final_df['Year of Graduation'] = year_of_grad

In [50]:
final_df['Year of Graduation'].value_counts()

Year of Graduation
0       1710
2021     430
2022     344
2020     151
2023     149
2024      21
2019       3
Name: count, dtype: int64

#### Replacing 0 as a Blank in 'Year of Graduation'

In [51]:
final_df['Year of Graduation'] = final_df['Year of Graduation'].replace({0 : ' '})

In [52]:
final_df.to_excel('Predicted Year of Graduation.xlsx')
final_df.to_csv('Predicted Year of Graduation.csv')